engine/src/witness/btc/smart_contract.rs

```diff
@@ -0,0 +1,338 @@
+	let bytes = script.as_script().as_bytes();
+
+	// First opcode must be OP_RETURN and second opcode must be OP_PUSHDATA1
+	if bytes[0] != OP_RETURN.to_u8() || bytes[1] != OP_PUSHDATA1.to_u8() {
+		return None;
+	}
+
+	// The next byte encodes the length of the remaining data blob:
+	let data_len = bytes[2];
+	let data_bytes = &bytes[3..];
+
+	// Sanity check:
+	if data_bytes.len() != data_len as usize {
+		return None;
+	}
+
+	Some(data_bytes)
+}
+
+#[allow(dead_code)]
+fn encode_swap_params_in_nulldata_utxo(params: UtxoEncodedData) -> Option<ScriptBuf> {
+	encode_data_in_nulldata_utxo(&params.encode())
+}
+
+fn script_buf_to_script_pubkey(script: &ScriptBuf) -> Option<ScriptPubkey> {
+	fn data_from_script<const LEN: usize>(script: &ScriptBuf, bytes_to_skip: usize) -> [u8; LEN] {
+		script.bytes().skip(bytes_to_skip).take(LEN).collect_vec().as_array()
+	}
+
+	let pubkey = if script.is_p2pkh() {
+		ScriptPubkey::P2PKH(data_from_script(script, 3))
+	} else if script.is_p2sh() {
+		ScriptPubkey::P2SH(data_from_script(script, 2))
+	} else if script.is_v1_p2tr() {
+		ScriptPubkey::Taproot(data_from_script(script, 2))
+	} else if script.is_v0_p2wsh() {
+		ScriptPubkey::P2WSH(data_from_script(script, 2))
+	} else if script.is_v0_p2wpkh() {
+		ScriptPubkey::P2WPKH(data_from_script(script, 2))
+	} else {
+		ScriptPubkey::OtherSegwit {
+			version: script.witness_version()?.to_num(),
+			program: script.bytes().skip(2).collect_vec().try_into().ok()?,
+		}
+	};
+
+	Some(pubkey)
+}
+
+// Currently unused, but will be used by the deposit wintesser:
+#[allow(dead_code)]
+pub fn try_extract_contract_call(
+	tx: &VerboseTransaction,
+	vault_address: ScriptPubkey,
+) -> Option<BtcContractCall> {
+	// A correctly constructed transaction carrying CF swap parameters must have at least 3 outputs:
+	let [utxo_to_vault, nulldata_utxo, change_utxo, ..] = &tx.vout[..] else {
+		return None;
+	};
+
+	// First output must be a deposit into our vault:
+	if utxo_to_vault.script_pubkey.as_bytes() != vault_address.bytes() {
+		return None;
+	}
+
+	// Second output must be a nulldata UTXO (with 0 amount):
+	if nulldata_utxo.value.to_sat() != 0 {
+		return None;
+	}
+
+	let mut data = try_extract_utxo_encoded_data(&nulldata_utxo.script_pubkey)?;
+
+	let Ok(data) = UtxoEncodedData::decode(&mut data) else {
+		tracing::warn!("Failed to decode UTXO encoded data targeting our vault");
+		return None;
+	};
+
+	// Third output must be a "change utxo" whose address we assume to also be the refund address:
+	let Some(refund_address) = script_buf_to_script_pubkey(&change_utxo.script_pubkey) else {
+		tracing::error!("Failed to extract refund address");
+		return None;
+	};
+
+	let deposit_amount = utxo_to_vault.value.to_sat();
+
+	// Derive min price (encoded as min output amount to save space):
+	let min_price = sqrt_price_to_price(bounded_sqrt_price(
+		data.parameters.min_output_amount.into(),
+		deposit_amount.into(),
+	));
```

Comment:
We should make it possible to pass the minimum_amount instead of price when witnessing the contract swaps. The price was used as a workaround for channels, because channels are opened before the swap is made. That's not the case with contract swaps. All we are doing here is converting to a price, which is then converted back to a min output on-chain.

Reply:
It is not clear that we should prefer the minimum amount. If we use DCA, for example, we want to know the price rather than the amount since we have to derive the min output for each chunk individually. Same for CCM. I'd rather we use Price consistently when requesting a swap from the swapping pallet, and think of the conversion to/from min output here simply as a way to compress this parameter in contract calls.

Reply:
Ok, makes sense 👌